rfcs/0013-ai-policies.md

```diff
@@ -0,0 +1,542 @@
+
+This RFC proposes a new Kuadrant API in the general AI Gateway realm:
+
+- `TokenRateLimitPolicy`
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+AI workloads are becoming extremely popular. Our policies can augment and enhance both the management of ingress traffic via the Gateway API and the routing to running model servers.
+The aim is to augment, not replace, the [Gateway API Inference Extensions](https://gateway-api-inference-extension.sigs.k8s.io/) project.
+
+Using the [Inference Platform Admin](https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/002-api-proposal#inference-platform-admin) and [Inference Workload Owner](https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/002-api-proposal#inference-workload-owner) personas, some high-level use-cases we hope to enable are:
+
+- As an Inference Platform Admin, I want to protect my LLM infrastructure from being overwhelmed by limiting per user token usage across all models.
+- As an Inference Workload Owner, I want to apply token usage limits to specific models so that costs are controlled based on the heaviest GPU using models.
+- As an Inference Workload Owner, I want to apply different token usage limits to different user groups based on business requirements.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New Policies
+
+### `TokenRateLimitPolicy`
+
+A Kuadrant `TokenRateLimitPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`) designed to allow users to define and enforce rate limiting rules to control token usage within OpenAI-style LLM provider APIs. It allows users to set and enforce token budget constraints for Gateways, but also for individual services exposed as HTTPRoutes. Per-user or per-group token rate limiting can be enforced based on JWT claims.
+
+
+
+## Concepts Introduced
+
+- Token-based rate limiting: Controls usage by number of tokens, not just requests. Helps manage LLM costs.
+
+## Example Use Case
+
+Say you're serving a chat LLM behind a Gateway, which implements an OpenAI-style chat `/completion` API. You want:
+
+- Free-tier users limited to 20k tokens/day.
+
+You’d define the following policies and resources:
+
+> Notes (somewhere): OpenAI-style usage metrics in responses required: `"usage":{"prompt_tokens":5,"total_tokens":15,"completion_tokens":10}`
+
+```yaml
+apiVersion: kuadrant.io/v1
+kind: AuthPolicy
+metadata:
+  name: gold-users-auth
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  rules:
+    authorization:
+      "gold-users":
+        opa:
+          rego: |
+            groups := split(object.get(input.auth.identity.metadata.annotations, "kuadrant.io/groups", ""), ",")
+            allow { groups[_] == "gold" }
+---
```

Comment:
I get this is just an example but multiple authpolicies here seems odd? trying to recall if multiple policies targeting the same resource even work

Reply:
i'll change this - was written before impl, the examples from the tutorials do similar with a single policy